File: selection/rust/src/error_manager.rs
use std::fmt::Display;

use crate::error_manager_configurations::ErrorManagerConfigurations;

#[derive(Debug, Default, Clone)]
pub struct ErrorManager {
    pub configurations: ErrorManagerConfigurations,
}

impl ErrorManager {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn setup(&mut self, configurations: ErrorManagerConfigurations) -> bool {
        if self.configurations.base.is_disabled_for_handling(false) {
            return false;
        }

        self.configurations = configurations;

        true
    }

    pub fn reset(&mut self) -> bool {
        if self.configurations.base.is_disabled_for_handling(false) {
            return false;
        }

        self.configurations = ErrorManagerConfigurations::default();

        true
    }

    pub fn transform_to_error_log<T>(&self, message: T) -> String
    where
        T: Display,
    {
        format!(
            "{}{}{}",
            self.configurations.title, self.configurations.title_message_separator, message,
        )
    }

    pub fn transform_to_error_log_with_title<TTitle, TMessage>(
        &self,
        title: TTitle,
        message: TMessage,
    ) -> String
    where
        TTitle: Display,
        TMessage: Display,
    {
        format!(
            "{}{}{}",
            title, self.configurations.title_message_separator, message,
        )
    }

    pub fn handle_error_output_conditions(&self, error_log: &str) -> Result<(), String> {
        if !self
            .configurations
            .base
            .is_enabled_for_runtime_execution_handling()
        {
            return Ok(());
        }

        if self.configurations.is_enabled_for_console_output() {
            println!("{error_log}");
        }

        if self.configurations.is_enabled_for_runtime_throw_output() {
            return Err(error_log.to_owned());
        }

        Ok(())
    }

    pub fn handle<T>(&self, message: T) -> Result<(), String>
    where
        T: Display,
    {
        let error_log = self.transform_to_error_log(message);

        self.handle_error_output_conditions(&error_log)
    }

    pub fn handle_with_title<TTitle, TMessage>(
        &self,
        title: TTitle,
        message: TMessage,
    ) -> Result<(), String>
    where
        TTitle: Display,
        TMessage: Display,
    {
        let error_log = self.transform_to_error_log_with_title(title, message);

        self.handle_error_output_conditions(&error_log)
    }
}

File: selection/python/qlogicae_cor/v1/error_manager.py
from qlogicae_cor.v1.error_manager_configurations import (
    ErrorManagerConfigurations,
)


class ErrorManager:
    def __init__(self) -> None:
        self._configurations = ErrorManagerConfigurations()

    def setup(
        self,
        new_configurations: ErrorManagerConfigurations,
    ) -> bool:
        try:
            if self._configurations.is_disabled_for_handling(
                new_configurations is None,
            ):
                return False

            self._configurations = new_configurations

            return True

        except Exception as exception:
            self.handle_error_outputs(exception)

            return False

    def reset(
        self,
    ) -> bool:
        try:
            if self._configurations.is_disabled_for_handling():
                return False

            self._configurations = ErrorManagerConfigurations()

            return True

        except Exception as exception:
            self.handle_error_outputs(exception)

            return False

    def transform_to_error_log(
        self,
        value1: str | Exception,
        value2: str | None = None,
    ) -> str:
        if isinstance(value1, Exception):
            return (
                f"{type(value1).__name__}"
                f"{self._configurations.title_message_separator}"
                f"{value1}"
            )

        if value2 is not None:
            return f"{value1}{self._configurations.title_message_separator}{value2}"

        return (
            f"{self._configurations.title}"
            f"{self._configurations.title_message_separator}"
            f"{value1}"
        )

    def handle_error_outputs(
        self,
        error: str | Exception,
        message: str | None = None,
    ) -> bool:
        if isinstance(error, Exception):
            self.handle_error_output_conditions(
                self.transform_to_error_log(error),
            )

            return True

        if message is not None:
            self.handle_error_output_conditions(
                self.transform_to_error_log(
                    error,
                    message,
                ),
            )

            return True

        self.handle_error_output_conditions(
            self.transform_to_error_log(error),
        )

        return True

    def handle_error_output_conditions(
        self,
        error_log: str,
    ) -> bool:
        if not self._configurations.is_enabled_for_runtime_execution_handling():
            return False

        if self._configurations.is_enabled_for_console_output():
            print(error_log)

        if self._configurations.is_enabled_for_runtime_throw_output():
            raise RuntimeError(error_log)

        return True
